lieer/labels_translation.py

```diff
@@ -0,0 +1,332 @@
+  that should by synced to Gmail;
+
+- Printing detailed information regarding label translation and
+  filtering;
+
+Usage:
+
+After creating an instance of LabelTranslator, it is necessary to call
+the instance's `load_user_translation` method with an argument that is
+a path of a file containing a translation map and filter patterns
+definitions.
+
+This file should be a JSON file having the following structure:
+
+{
+    "labels_map": {
+        "<remote label>" : "<local tag>",
+        "<remote label>" : "<local tag>",
+        .
+        .
+        .
+    },
+    "ignore_patterns": [
+        "<pattern>",
+        "<pattern>",
+        .
+        .
+        .
+    ],
+    "label_sep": "<label separator>"
+}
+
+Where:
+
+* "label_map" is a mapping between remote labels and local tags.
+
+* "ignore_patterns" is a list of regular expressions; local tags
+  matching any of these patterns will not be synced to Gmail.
+
+* "label_sep" is a string that will replace, in local tags, the slash
+  used by Gmail as a label/sublabel separator.
+
+For example. a custom translation file might look like this:
+
+{
+    "labels_map": {
+        "TRASH"     : "deleted",
+        "my-remote": "my-local"
+    },
+    "ignore_patterns": [
+        "new",
+        "par.*"
+    ],
+    "label_sep": "::"
+}
+
+Using this file will have the following effects:
+
+1. Gmail's 'TRASH' label will be translated to the local tag
+   'deleted' and vice versa;
+
+2. Gmail's label 'my-remote' will be translated to the local
+   tag 'my-local' and vice versa;
+
+3. The local tag 'new' will not be synced to Gmail;
+
+4. Any local tag matching the regular expression 'par.*' (i.e., any
+   tag that starts with 'par') will not be synced to Gmail;
+
+5. The '/' character that separates label/sublabel in Gmail will be
+   replaced, in local tags, with the string '::'. For example, a
+   Gmail 'foo/bar' label will be translated to the local tag 'foo::bar'.
+"""
+
+import os
+import json
+import re
+import notmuch
+
+class LabelTranslator:
+    """Translation between Gmail labels and Notmuch tags."""
+    
+    # the default separator between parent/child labels in Gmail
+    GMAIL_LABEL_SEP = '/'
+
+    # the default tranlation of label names between the local
+    # (notmuch) system and the remote (Gmail) system. This translation
+    # map can be overridden by a user's custom map.
+    DEFAULT_REMOTE_TO_LOCAL_MAP = {
+        'INBOX'     : 'inbox',
+        'SPAM'      : 'spam',
+        'TRASH'     : 'trash',
+        'UNREAD'    : 'unread',
+        'STARRED'   : 'flagged',
+        'IMPORTANT' : 'important',
+        'SENT'      : 'sent',
+        'DRAFT'     : 'draft',
+        'CHAT'      : 'chat'
+    }
+
+    # set of Gmail's system labels that should not be created by gmi.
+    # the labels in this set should be written in all lower case letters.
+    SYSTEM_IGNORED_LABELS = {
+        'attachment',
+        'encrypted',
+        'signed',
+        # 'new',
```

Comment:
I think new is set as new-tag for new notmuch configs. So I think it should also be ignored by default, but that should be configurable along with the rest of the custom tag-ignoring.

Reply:
According to the notmuch-config man page (on Debian), the default tags that are assigned to new mail are 'unread' and 'inbox'. It seems, however, that tagging new mail with 'new' is a common convention.

But since the choice of the tag(s) for new mail are dependant on conventions or user preferenc, in my opinion gmailieer should not hard-code it, and should let the user to decide. So my suggestion is to remove 'new' from the SYSTEM_IGNORED_LABLES list, and in case the user decide to use 'new' for tagging new mail, he or she could use the new filter mechanism to exclude this tag from syncing to GMail.

Reply:
Ok, sounds good. Lets keep 'new' outside the synced tags - but we then need to notify users who don't add it to their new label ignore list.